feat(puppeteer-core): Infer element type from complex selector

gomain opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. -->

**What kind of change does this PR introduce?**

Better type inference.

**Did you add tests for your changes?**

~Not yet.~ Yes.

**If relevant, did you update the documentation?**

Not yet.

**Summary**

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
<!-- Try to link to an open issue for more information. -->
Currently methods that return an element handle, i.e. `.$`, `.waitForSelector` attempt to infer the node element type from the selector string. However, this only works when the selector is an exact match of the element tag, i.e. a selector `"a"` would be inferred as `HTMLAnchorElement` . And not when the selector is complex, i.e. selectors `"a#some-id"`, `div > a`, `a:nth-child(2)` would all fallback on `Element`.

This is due to simply looking up the the selector in `HTMLElementTagNameMap` and `SVGElementTagNameMap` without any attempt to parse the selector string.

This PR is an attempt to do so.

**Does this PR introduce a breaking change?**

<!-- If this PR introduces a breaking change, please describe the impact and a migration path for existing applications. -->
This could break existing incorrect assertions using the `as` keyword.

**Other information**

~This PR introduces a dependency on the `type-fest` package.~

This PR is far from complete (no tests, no docs). Put out early for feedback and discussion.

gomain committed:
feat(puppeteer-core): infer element type from complex selector

OrKoN commented on packages/puppeteer-core/src/common/types.ts:
```diff
@@ -17,6 +17,7 @@
 import {JSHandle} from './JSHandle.js';
 import {ElementHandle} from './ElementHandle.js';
 import {LazyArg} from './LazyArg.js';
+import type {LastArrayElement, Split} from 'type-fest';
```
is it possible to implement two helpers that are needed instead of bringing in the entire lib?

gomain replied:
done.

gomain committed:
chore: refactor `LastArrayElement` as self implemented

gomain committed:
chore: refactor `Split` as self implemented

remove dependency on `type-fest`

gomain committed:
chore: refactor type evaluation style

Breakdown nested composition into intermediate steps. This is
preferred because typescript does not typecheck possibility of
intermediate nesting evaluating to non-constrainted types.

gomain committed:
chore: regenerate docs

gomain committed:
chore: strengthen `LastArrayElement`

Only accept non empty arrays.

gomain committed:
chore: rename `LastArrayElement` to just `Last`

gomain committed:
chore: remove eslint-disable-next-line

gomain committed:
chore: assert some resolve paths are `never` taken

gomain committed:
chore: resolve to `unknown` when fail

gomain committed:
chore: fix typo

gomain committed:
chore: test `NodeFor`

gomain committed:
chore: expand type tests for `.$` and `.$$`

gomain committed:
chore: regerate docs after fix typo

OrKoN committed:
Merge branch 'main' into infer-element-type-from-complex-selector
